src/Psalm/Type.php

```diff
@@ -769,12 +787,19 @@ private static function intersectAtomicTypes(
     /**
      * @psalm-assert-if-true TIterable|TNamedObject|TTemplateParam|TObjectWithProperties $type
```

Comment:
could you remove that assertion? Literal unions in `@psalm-assert` never worked correctly (there two incomplete PR open on that subject but none could get it to work properly yet. We probably should forbid this for the time being)

Reply:
I've half an idea on how to get unions working on all assertions, which would actually fit nicely with the changes of this MR which already deal with unions in assertions (even if just those generated from template types): this would require changing the Assertion API to work with Unions instead of atomics, and move the code I just wrote in this PR to the assertion reconcilers, might do in a future PRs.